webpack.config.js

```diff
@@ -21,46 +20,204 @@ module.exports = {
         "theme-light-custom": "./node_modules/matrix-react-sdk/res/themes/light-custom/css/light-custom.scss",
         "theme-dark-custom": "./node_modules/matrix-react-sdk/res/themes/dark-custom/css/dark-custom.scss",
     },
+
+    optimization: {
+        // Put all of our CSS into one useful place - this is needed for MiniCssExtractPlugin.
+        // Previously we used a different extraction plugin that did this magic for us, but
+        // now we need to consider that the CSS needs to be bundled up together.
+        splitChunks: {
+            cacheGroups: {
+                styles: {
+                    name: 'styles',
+                    test: /\.css$/,
+                    enforce: true,
+                    // Do not add `chunks: 'all'` here because you'll break the app entry point.
+                },
+            },
+        },
+
+        // Minification is normally enabled by default for webpack in production mode, but
+        // we use a CSS optimizer too and need to manage it ourselves.
+        minimize: argv.mode === 'production',
+        minimizer: argv.mode === 'production' ? [new TerserPlugin({}), new OptimizeCSSAssetsPlugin({})] : [],
+    },
+
+    resolve: {
+        // We define an alternative import path so we can safely use src/ across the react-sdk
+        // and js-sdk. We already import from src/ where possible to ensure our source maps are
+        // extremely accurate (and because we're capable of compiling the layers manually rather
+        // than relying on partially-mangled output from babel), though we do need to fix the
+        // package level import (stuff like `import {Thing} from "matrix-js-sdk"` for example).
+        // We can't use the aliasing down below to point at src/ because that'll fail to resolve
+        // the package.json for the dependency. Instead, we rely on the package.json of each
+        // layer to have our custom alternate fields to load things in the right order. These are
+        // the defaults of webpack prepended with `matrix_src_`.
+        mainFields: ['matrix_src_browser', 'matrix_src_main', 'browser', 'main'],
+        aliasFields: ['matrix_src_browser', 'browser'],
+
+        // We need to ensure we can resolve TS files, but that also means we need to define
+        // every single extension we might see, ever.
```

Comment:
Really? https://webpack.js.org/configuration/resolve/#resolveextensions claims it's only for when you're leaving the file extension off your import. Afaik we should only ever be doing this for `.js` and now `.ts` files. The docs also say the default is much smaller than this set so if we were relying on more than that, I don't know how it was working before.

Reply:
Empirical evidence of fonts going missing says otherwise. It's possible it was a different bug at the time, but I'm not totally convinced.

Reply:
Hmm, which fonts? Inconsolata and Twemoji are woff2, Nunito & Open Sans are ttf. Neither of those extensions are in the defaults so either the defaults have changed or there's something else going on here.

Reply:
All I know is I didn't have it specified and TypeScript didn't load. I defined it as `['.js', '.ts']` and suddenly the OS default fonts were being used. This isn't how it's supposed to work, but that one change broke it for reasons beyond comprehension.

Reply:
ok - I would suggest maybe putting it as `['.js', '.ts', '.ttf']` and commenting it so we can sleuth it down the line.

Reply:
well, it also breaks for roughly 50% of our images, some json files, and Olm.

Reply:
I guess it was a different bug at the time... bringing it down to js, ts, and json doesn't seem to have ill effects